File: c/src/classical/substitution/monoalphabetic/atbash.c
/**
 * @file atbash.c
 * @brief Atbash Cipher implementation
 */

#include "cryptology/classical/substitution/monoalphabetic/atbash.h"
#include "cryptology/alphabets.h"
#include <string.h>
#include <ctype.h>

#define DEFAULT_ALPHABET ENGLISH_ALPHABET

static int find_char_in_alphabet(char c, const char *alphabet) {
    const char *pos = strchr(alphabet, c);
    return pos ? (int)(pos - alphabet) : -1;
}

int atbash_encrypt(const char *plaintext, const char *alphabet,
                   char *result, size_t result_size) {
    if (!plaintext || !result || result_size == 0) {
        return -1;
    }
    
    if (!alphabet) {
        alphabet = DEFAULT_ALPHABET;
    }
    
    size_t alphabet_len = strlen(alphabet);
    size_t plaintext_len = strlen(plaintext);
    
    if (plaintext_len >= result_size) {
        return -1;
    }
    
    for (size_t i = 0; i < plaintext_len; i++) {
        char c = tolower((unsigned char)plaintext[i]);
        int pos = find_char_in_alphabet(c, alphabet);
        
        if (pos != -1) {
            // Character is in alphabet - reverse position
            int reversed_pos = alphabet_len - 1 - pos;
            result[i] = alphabet[reversed_pos];
        } else {
            // Character not in alphabet - keep unchanged
            result[i] = plaintext[i];
        }
    }
    
    result[plaintext_len] = '\0';
    return 0;
}

int atbash_decrypt(const char *ciphertext, const char *alphabet,
                   char *result, size_t result_size) {
    // Atbash is symmetric - encrypt and decrypt are the same
    return atbash_encrypt(ciphertext, alphabet, result, result_size);
}

int atbash_produce_alphabet(const char *alphabet, char *result, size_t result_size) {
    if (!alphabet || !result || result_size == 0) {
        return -1;
    }
    
    size_t alphabet_len = strlen(alphabet);
    if (alphabet_len >= result_size) {
        return -1;
    }
    
    // Reverse the alphabet
    for (size_t i = 0; i < alphabet_len; i++) {
        result[i] = alphabet[alphabet_len - 1 - i];
    }
    
    result[alphabet_len] = '\0';
    return 0;
}

File: python/cryptology/classical/substitution/monoalphabetic/atbash.py
"""
Atbash Cipher implementation.

The Atbash cipher is a monoalphabetic substitution cipher where the alphabet is reversed.
The first letter is replaced with the last, the second with the second-to-last, and so on.
Like ROT13, it is symmetric - encrypting twice returns the original text.
"""

# Default English alphabet (lowercase only)
DEFAULT_ALPHABET = "abcdefghijklmnopqrstuvwxyz"


def encrypt(plaintext: str, alphabet: str = DEFAULT_ALPHABET) -> str:
    """
    Encrypt plaintext using Atbash cipher.
    
    Args:
        plaintext: The text to encrypt (will be converted to lowercase)
        alphabet: The alphabet to use for encryption (default: English lowercase)
    
    Returns:
        The encrypted ciphertext
    
    Example:
        >>> encrypt("HELLO")
        'svool'
    """
    # Convert input to lowercase
    plaintext = plaintext.lower()
    result = []
    
    for char in plaintext:
        if char in alphabet:
            # Find position in alphabet
            pos = alphabet.index(char)
            # Reverse position: first <-> last, second <-> second-to-last, etc.
            reversed_pos = len(alphabet) - 1 - pos
            result.append(alphabet[reversed_pos])
        else:
            # Keep characters not in alphabet unchanged
            result.append(char)
    
    return ''.join(result)


def decrypt(ciphertext: str, alphabet: str = DEFAULT_ALPHABET) -> str:
    """
    Decrypt ciphertext using Atbash cipher.
    
    Since Atbash reverses the alphabet, encrypting twice returns
    the original text. Therefore, decrypt is the same as encrypt.
    
    Args:
        ciphertext: The text to decrypt (will be converted to lowercase)
        alphabet: The alphabet that was used for encryption (default: English lowercase)
    
    Returns:
        The decrypted plaintext
    
    Example:
        >>> decrypt("svool")
        'hello'
    """
    return encrypt(ciphertext, alphabet)


def produce_alphabet(alphabet: str = DEFAULT_ALPHABET) -> str:
    """
    Produce an Atbash-reversed alphabet.
    
    This method creates a custom alphabet by reversing the base alphabet.
    The produced alphabet can be used with polygraphic ciphers for enhanced security.
    
    Args:
        alphabet: The base alphabet to reverse (default: English lowercase)
    
    Returns:
        An Atbash-reversed alphabet
    
    Example:
        >>> produce_alphabet()
        'zyxwvutsrqponmlkjihgfedcba'
        >>> produce_alphabet("abc")
        'cba'
    """
    return alphabet[::-1]

